### ISSUE TITLE
Questions re .pyc cleanup step

### ISSUE THREAD

**Issue by lazka** [flatpak/flatpak-builder#247]
## Linux distribution and version

Debian sid

## Flatpak-builder version

flatpak-builder 1.0.1

## Description of the problem

As things are now flatpak-builder removes all .pyc files that get created during the build with distutils/setuptools. Example: https://flathub.org/builds/#/builders/16/builds/721/steps/5/logs/stdio
Looking at the code it looks like it shouldn't, but it also hardcodes the mtime to bytes 4-8 in fixup_python_time_stamp() which is no longer the case with Python 3.7, so not sure.

A related issue for the fdo-sdk is discussed here: https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues/554 where the stdlib .pyc files have a wrong mtime. The current consensus there is to move Python to hash based verification (instead of mtime) by default which is supported since Python 3.7.

So, assuming this gets solved in the fdo-sdk, and the current broken state of the pyc cleanup code I'm wondering if it shouldn't be removed.

In case someone wonders why having this working would be useful: For Quod Libet having .pyc for its own source files (excluding stdlib) saves 500ms on start.

**Comment by nanonyme** [comment]
I don't think it in general can be removed but it would be nice if we could gradually deprecate and kill the functionality through opt-out. FD.o 1.6 has Python 2 in Platform, FD.o 18.08 and derivatives has Python 2 only in SDK and FD.o 19.08 will not have Python 2 at all.

**Comment by nanonyme** [comment]
Note that flatpak-builder should most likely set SOURCE_DATE_EPOCH to something reasonable so our Python 3.7 uses CHECKED_HASH bytecode by default.

**Comment by nanonyme** [comment]
In runtime we're hardcoding it into a constant like this https://buildstream.gitlab.io/buildstream/format_project.html#builtin-defaults

**Comment by nanonyme** [comment]
Magic byte contains Python version. It might be best if this entire logic was disabled on 3.7 and newer. Ping @alexlarsson

**Comment by alexlarsson** [comment]
The mtime munging is definately needed on older versions, because ostree itself will modify the mtime of things, so the embedded mtimes will be wrong. However, if python changed we need to update this.

**Comment by nanonyme** [comment]
@alexlarsson the core thing here is format changed and bytecode may or may not have timestamps at all so you can't simply write without parsing the header

**Comment by alexlarsson** [comment]
Yeah, the code needs updating. But we can't just disable it.

**Comment by nanonyme** [comment]
In timestamped mode yes you still need to reset

**Comment by nanonyme** [comment]
There's some more elaborate bytecode analysis in https://gitlab.com/BuildStream/buildstream/issues/836

**Comment by nanonyme** [comment]
FWIW how we handle the timestamp in runtime is we just hardcode it to integer value of 2011-11-11.

**Comment by alexlarsson** [comment]
The spec for the new header is here: https://www.python.org/dev/peps/pep-0552/#id15

**Comment by alexlarsson** [comment]
And Python 3.7.2 has importlib.util.MAGIC_NUMBER= b'B\r\r\n'

**Comment by alexlarsson** [comment]
While 3.5.2 has importlib.util.MAGIC_NUMBER=b'\x16\r\r\n'

**Comment by alexlarsson** [comment]
Here is the full list: https://github.com/python/cpython/blob/9e14e49f13ef1a726f31efe6689285463332db6e/Lib/importlib/_bootstrap_external.py#L167

**Comment by nanonyme** [comment]
So 3393?

**Comment by alexlarsson** [comment]
So, #256 fixes the header rewrite, and handles the new hash-based verification format.
I also merged a change that makes f-b set SOURCE_DATE_EPOCH which causes py3.7 to use the hash-based pyc headers. Unfortunately that seems to trigger a bug in QuodLibet:

```
$ flatpak build app
sh-4.4$ /app/bin/quodlibet 
E: 0.157: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.157: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.157: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.158: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.158: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.158: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.158: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.159: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.159: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.159: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.160: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.160: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.160: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.161: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.161: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.161: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
E: 0.162: util.importhelper.load_dir_modules: <frozen importlib._bootstrap_external>:838:get_code: TypeError: a bytes-like object is required, not 'str'
```

**Comment by alexlarsson** [comment]
This seems to be some version of https://bugs.python.org/issue34056
@valentindavid This is claimed to be fixed in python 3.7.1 RC1, maybe we should update the runtime to 3.7.2?

**Comment by alexlarsson** [comment]
@lazka see override info in https://github.com/flatpak/flatpak-builder/pull/257

**Comment by nanonyme** [comment]
FWIW https://gitlab.com/freedesktop-sdk/freedesktop-sdk/blob/18.08/elements/base/python3.bst#L68

**Comment by nanonyme** [comment]
18.08 should have the fix already (it's pre-3.7.1) and runtime was imported. Gnome runtime probably does not yet

**Comment by alatiera** [comment]
Gnome 3.30 currently is based upon fd.o 18.08.22 I believe which means it still has `python 3.7.0`. We just bumped the Nigthly runtime to .25 junction of fd.o, will bump the stable runtime as well and trigger an import.
